vmm/src/lib.rs

```diff
@@ -2080,6 +2089,10 @@ impl Vmm {
             UpdateNetworkInterface(netif_update, sender) => {
                 Vmm::send_response(self.update_net_device(netif_update), sender);
             }
+            ConfigureVm(vmm_config, sender) => match self.configure_from_json(*vmm_config) {
+                Err(e) => Vmm::send_response(Err(e), sender),
+                _ => Vmm::send_response(self.start_microvm(), sender),
```

Comment:
I suggest to implement an exit routine for the case when the configure_from_json() fails to avoid leaving the vmm in a partially configured state. 
Also an integration for this to check we keep this behavior.

Reply:
I've added in the commit a solution for the first one. Please, take a look.

Reply:
I think rolling back to the original configuration is too complex, and IIRC the `StartMicroVm` operation can already leave the VM(M) in an inconsistent state when something fails. Exiting the process is an option, but it's strange to have some API calls exit on error, while others don't. 

Ultimately I think the semantics of this new API call should keep as close as possible to sending the individual requests one by one. @sandreim has a very fair point that right now the configuration order is opaque, and the user cannot really infer which operations were successful looking at the error message. IMO, the cleanest solution here is to guarantee the resources will be configured in a certain order as long as they are present, so it's relatively easy to determine the current state based on which configuration failed in particular. If our current error messages are not informative enough, I would simply go with the more specific error message option, without any other changes.

Reply:
> the StartMicroVm operation can already leave the VM(M) in an inconsistent state when something fails

I think this should not be used as an argument for any decision. This is in itself a bug/issue that needs to be fixed. If we decide to exit on irrecoverable errors, we can also exit here - in which case I personally prefer the delayed exit so that an API response can be delivered.

>  IMO, the cleanest solution here is to guarantee the resources will be configured in a certain order as long as they are present, so it's relatively easy to determine the current state based on which configuration failed in particular. If our current error messages are not informative enough, I would simply go with the more specific error message option, without any other changes.

Looking at the code, this would be harder and bring more complexity than a simple rollback:

1. clean sleight - just purge all existing configs on errors - only allow single API call if no other individual resource config api calls happened (it's right there in the name: **single** api call).
or,
2. work with a mirror/backup of the configs and only commit changes if everything went well.

> so it's relatively easy to determine the current state based on which configuration failed in particular.

This is very complex for the user, they'd likely just kill and start another Firecracker than recompose a partial single API with only the failing bits.

--------------

In the case of irrecoverable errors, I vote for sending the user a farewell message and committing seppuku.

Reply:
I am wondering ...why do we want to provide a single API call for configuring the microvm **and** starting it? Why not provide an API call only for configuring the microVM?
Even the phrasing is weird IMO: you have an API call called vm-config that configures **and** starts the microvm? 
I am proposing this:
* This PR implements support for configuring the microVM (and only that) -> i.e. you cannot issue and action requests with this API call since its purpose is configuration of the microVM.
* Users can call into this any time they want (IMO this has all the logic in the world since users can issue a PUT request on any resource (except for the logger) any number of times before starting a microVM), right?
* Change the logic to not receive an error when a user wants to reinitialize the logger; Come to think of it, why would we punish the user with an error if by mistake he/she issues the logger  initialization a second time?

Reply:
And yes,  yes,`start_microvm` can leave the vm in an inconsistent state and we definitely have to fix that (looked for an issue tracking this but have not found one :( ); however I think that is a separate issue with a separate PR and it should not impact our decision here.
@lauralt can you please check if we have an issue for this problem and if not open one for it?

Reply:
> I am wondering ...why do we want to provide a single API call for configuring the microvm **and** starting it? Why not provide an API call only for configuring the microVM?
> Even the phrasing is weird IMO: you have an API call called vm-config that configures **and** starts the microvm?
> I am proposing this:
> 
>     * This PR implements support for configuring the microVM (and only that) -> i.e. you cannot issue and action requests with this API call since its purpose is configuration of the microVM.

This is a fairly good point, but maybe the behavior is intended as a one shot start by using the API the same way it is possible with passing a json file on the command line.
I think it is a good idea to rename this API endpoint to something like` /start-from-config` or anything that provides a complete and clear expectation of what it does.

I would also vote for providing a clear response to the user (5xx) and exiting the process.

Reply:
Cool, let's have a chat and get back with the conclusion!

Reply:
I've added another solution, please take a look! 
@dianpopa  I would leave the request path name as it is because I don't want to make it more complicated. It is mentioned in the documentation and swagger file that this API call also starts the microvm, so I think/hope it's pretty clear what it does. I would also not separate the configuration of the vm from the start action because, as @sandreim said, I think we preffer an one shot start. 
I've made the error message for the bad configuration case more explicit and, because we can't be sure of the `thread::sleep` behaviour, I'd suggest to not exit the process.

Reply:
@lauralt I was not suggesting any renaming of the request path name, or was I? what path name exactly?
I was not aware that we want a **one shot start**, I was more under the impression that we want to implement a **one shot vm config**. This feature request was intended as a solution to the fact that there is non-negligible time spent in API calls by some of our customers and we would like to do our best in that way while also keeping consistency with the old behavior. By providing a one shot start, we are actually limiting the usability of our product. 
We are forcing an `InstanceStart` after the user has configured the vm. I have a question: what if the user wants to trigger the `FlushMetrics` action right before the vm is started and then again after the vm started, how do you do that with the **one shot start**?

Reply:
> We are forcing an InstanceStart after the user has configured the vm. I have a question: what if the user wants to trigger the FlushMetrics action right before the vm is started and then again after the vm started, how do you do that with the one shot start?

We are making a conscious compromise here, we are compromising _manoeuvrability_ for _speed_:

- If the user wants tight control of ordering of API operations they can use the individual specific API calls in whichever order they want.
- If the user wants a quick start they can use the single-api-call.

I vote for keeping configuring and starting as part of the same operation because _speed_ is what we're aiming with this option.

> @dianpopa I would leave the request path name as it is because I don't want to make it more complicated.

@lauralt I, too, would prefer a change in the request path name to also include 'start' in the URL.
It's a simple rename of a newly added path, there is no "complication" involved in renaming it.

Reply:
> > We are forcing an InstanceStart after the user has configured the vm. I have a question: what if the user wants to trigger the FlushMetrics action right before the vm is started and then again after the vm started, how do you do that with the one shot start?
> 
> We are making a conscious compromise here, we are compromising _manoeuvrability_ for _speed_:
> 
> * If the user wants tight control of ordering of API operations they can use the individual specific API calls in whichever order they want.
> * If the user wants a quick start they can use the single-api-call.
> 
> I vote for keeping configuring and starting as part of the same operation because _speed_ is what we're aiming with this option.

I hope you realize that "speed" here means the period it takes for  a `start_instance` API call to finish (a little less than that actually). If you do realize that and still think that is worth compromising _maneuverability_ and thus restricting the group of users that can use this feature,  I will be happy to disagree and commit.

> > @dianpopa I would leave the request path name as it is because I don't want to make it more complicated.
> 
> @lauralt I, too, would prefer a change in the request path name to also include 'start' in the URL.
> It's a simple rename of a newly added path, there is no "complication" involved in renaming it.